runtime/src/chplcast.c

```diff
@@ -64,72 +65,69 @@ static int illegalFirstUnsChar(char c) {
       str++;                                                            \
-    numitems = sscanf(str, format"%n", &val, &numbytes);                \
-    if (scanningNCounts() && numitems == 2) {                           \
-      numitems = 1;                                                     \
     }                                                                   \
-    if (numitems == 1) {                                                \
-      while(str[numbytes] && isspace(str[numbytes]))                    \
-        numbytes++;                                                     \
-      if (numbytes == strlen(str)) {                                    \
-        /* for negatives, sscanf works, but we wouldn't want chapel to */ \
-        if (uns && illegalFirstUnsChar(*str)) {                         \
-          *invalid = 1;                                                 \
-          *invalidCh = *str;                                            \
-        } else {                                                        \
-          *invalid = 0;                                                 \
-          *invalidCh = '\0';                                            \
-        }                                                               \
-      } else {                                                          \
-        *invalid = 1;                                                   \
-        *invalidCh = *(str+numbytes);                                   \
+    if (strlen(str) >= 2 && str[0] == '0') {                            \
+      if (str[1] == 'b' || str[1] == 'B') {                             \
+        numberBase = 2;                                                 \
+        str += 2;                                                       \
+      } else if (str[1] == 'o' || str[1] == 'O') {                      \
+        numberBase = 8;                                                 \
+        str += 2;                                                       \
+      } else if (str[1] == 'x' || str[1] == 'X') {                      \
+        numberBase = 16;                                                \
+        str += 2;                                                       \
       }                                                                 \
+    }                                                                   \
+    if (str[0] == '-' || str[0] == '+') {                               \
+      *invalid = 1;                                                     \
+      *invalidCh = *str;                                                \
+      return -1;                                                        \
+    }                                                                   \
+    if (negative) {                                                     \
+      newStr = chpl_malloc(strlen(str) + 2);                            \
```

Comment:
We typically try to use the `chpl_mem_*` api (`chpl_mem_alloc`/`chpl_mem_free`) instead of calling down to the specific implementation when possible. The main difference is that it enables leak and memory tracking. 

I also worry about the added cost of an allocation for string casting. Can you check the performance of test/types/string/cassella/cast-from-string.chpl?

Reply:
FYI here's what I see on chapcs07:

| branch | int64 | int32 | int16 | int8  |
| ------ | ----- | ----- | ----- | ----- |
| before | 1.36s | 0.32s | 0.29s | 0.28s |
| after  | 0.61s | 0.52s | 0.43s | 0.41s |

~1.5x slowdown for 8, 16, and 32 bit casts (presumably from allocation)

~2x speedup for 64 bit casts (guessing because of switching from scanf to strtoll)

Reply:
Thanks @ronawho.  I'm switching to chpl_mem_alloc and will take a look at the performance a bit myself too.  I think it's a net positive if your chapcs07 numbers are representative in general since int(64) will be the majority of casts in practice.

Reply:
I agree that it's a net positive, though presumably if you could eliminate the allocation the 64 bit numbers would be even better.